### ISSUE TITLE
Need to handle server connectivity changes

### ISSUE THREAD

**Issue by lukehutch** [serverpod/serverpod#764]
I want my app to notify the user if the Serverpod server is not responding, and also to keep polling the server until it can be reached.

I tried using `FlutterConnectivityMonitor` for this, but it only notifies about the link status being up or down, not whether the server is reachable.

In `SessionManager.initialize()`, `refreshSession()` is called only once. If that fails, the user is not notified, and no further attempt is made to connect.

There should be an attempt to keep retrying `refreshSession()` every few seconds until a connection is established. And if the connection drops at any point, via any connection attempt made through the `Client`, then the polling to check when the server comes back online should resume.

There also needs to be a listener mechanism for listening to changes in server online status.

**Comment by lukehutch** [comment]
#764 returns `bool` from `SessionManager.initialize()`, which at least tells the client whether a session could be established when the first attempt is made. However there needs to be a proper connection watchdog put in place that connects and reconnects to the server as necessary, and notifies listeners of connectivity changes.

**Comment by lukehutch** [comment]
I added #1012, which adds `onFailedCall` and `onSucceededCall` to `ServerpodClient`, so that user code can better track connectivity status.

**Comment by lukehutch** [comment]
By these arguments, you wouldn't need `onSucceededCall` either! But it must be used for something, or you wouldn't have added it. By symmetry there should be an `onFailedCall` too.

I want a central way in my app to notify the user (via a `SnackBar` or similar) that the server is not currently reachable. I can wrap every server call in boilerplate `try...catch` logic, and repeat the same `SnackBar` code in every case, or I can have the server notify my code in one central place that a call failed with a specific error object.

I have a fairly complex state machine that handles routing state for the whole app. This shows an interstitial loading screen while the server is being reached (i.e. while `SessionManager` is being initialized), while the initial data is being fetched from the server, and it also will not let the user use the app if certain permissions are not granted. I was using `onFailedCall` to show the interstitial if the server was unreachable, since my app is unusable in that case (although this is probably too heavy-handed, so now I just show `SnackBar`s instead).

**Comment by lukehutch** [comment]
By the way the Pixorama connectivity detection does not actually work. #1687

**Comment by SandPod** [comment]
> By these arguments, you wouldn't need `onSucceededCall` either! But it must be used for something, or you wouldn't have added it. 

That was my thinking as well, and we don't have `onSuccededCall` in the project yet as far as I understand since it is also added in the PR.

> I want a central way in my app to notify the user (via a SnackBar or similar) that the server is not currently reachable....

 OK, I think I have a better understanding now and the requirements make sense, to have some centralized way of handling a non-responsive server.
 
One thing that puzzles me with the current implementation is that the exception will still be thrown if the call fails. That should result in you having to wrap every server call in a `try...catch` block to prevent the exception from propagating.

How is that currently solved given your PR?

The PR approach will also target every single call to the Serverpod server. Is the error handling really shared for all calls to the server? My thinking is that there might be a need for different things to happen depending on what method has an error.

An alternative solution would be to group the server call methods in some business logic that applies the same `try...catch` logic.

**Comment by lukehutch** [comment]
Oops, you're right, I wrote this so long ago that I forgot that I added `onSucceededCall` too, somehow I mis-remembered that was already part of Serverpod!

But yes, this is about tracking server reachability (`onSucceededCall`) and server errors (`onFailedCall`), in a central way.

N.B. server reachability is not the same thing as network connectivity, or even server IP visibility, because the server process might be down, or something.
